support for multiline regexp delimited messages

serii833 opened the pull request.

serii833 committed:
support for multiline regexp delimited messages

serii833 committed:
fix for multiline delimiter

arnecls commented on tio/bufferedreader.go:
```diff
@@ -89,15 +95,16 @@ var BufferDataInvalid = bufferError("Invalid data")
+	encoding     binary.ByteOrder
+	flags        BufferedReaderFlags
+	incomplete   bool
+	delimiterStr string
```
instead of storing the string here, store the compiled regex, so regex compiling is done once during initialisation.

arnecls commented on tio/bufferedreader.go:
```diff
@@ -111,22 +118,30 @@ type BufferedReader struct {
 
+	
+	if flags == BufferedReaderFlagMultiline {
+		buffer.parse = buffer.parseDelimiterMultiLine
```
do the regex compiling here

arnecls commented on tio/bufferedreader.go:
```diff
@@ -257,6 +272,55 @@ func (buffer *BufferedReader) parseMLE64() ([]byte, int) {
 
+// messages are separated by regexp
+func (buffer *BufferedReader) parseDelimiterMultiLine() ([]byte, int) {
+	startRegexp := regexp.MustCompile(buffer.delimiterStr)
```
move this to the init function and store the compiled regex for improved performance (see above)

arnecls commented on tio/bufferedreader.go:
```diff
@@ -257,6 +272,55 @@ func (buffer *BufferedReader) parseMLE64() ([]byte, int) {
+	loopStartIdx := 0
+	endIdx := 0
+
+	for true {
```
for {

arnecls commented on tio/bufferedreader.go:
```diff
@@ -257,6 +272,55 @@ func (buffer *BufferedReader) parseMLE64() ([]byte, int) {
+	endIdx := 0
+
+	for true {
+		endIdx = bytes.Index(buffer.data[loopStartIdx:buffer.end], lineDelimiter)
```
golang regexp has multiline support.
See https://github.com/google/re2/wiki/Syntax and https://github.com/trivago/gollum/issues/228

serii833 replied:
I have tried multiline regexp. but in case of using regexp last message did not get captured until next message arrives. 
this implementation will capture last message

arnecls reviewed (changes requested):
Thanks for the PR.
There's a few improvements that should be made, but in general I think this is a good addition.

serii833 committed:
code review changes

arnecls commented on tio/bufferedreader.go:
```diff
@@ -257,6 +271,55 @@ func (buffer *BufferedReader) parseMLE64() ([]byte, int) {
+		endIdx++
+		endIdx += loopStartIdx
+
+		startLoc := startRegexp.FindIndex(buffer.data[loopStartIdx:endIdx])
```
I finally got some time to do an in depth review.
What I'm asking myself is - why do you even go line-by-line here?
You could just start looking for the regexp in the complete message chunk, which is just a stream of bytes containing linefeeds. Using FindIndex will ignore the linefeeds anyways and you could even use expressions spanning multiple lines this way.
See https://play.golang.org/p/egztEi8m7am

Also this would drastically simplify this loop.

arnecls replied:
See my comment on the unit test why it might not be a good idea to treat trailing new lines as a valid message.

arnecls commented on tio/bufferedreader_test.go:
```diff
@@ -54,6 +55,34 @@ func TestBufferedReaderDelimiter(t *testing.T) {
+			"  at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:676)\n" +
+			"  at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:190)\n" +
+			"  ... 21 common frames omitted\n",
+			"3333-12-06 3\n",
```
In this example - how do you know that the last message is complete without the start of the next message?
In a network or large message scenario it might be that the message is cut in the middle and you have to continue reading. If you don't do this you will end up with two message instead of one.

serii833 replied:
I agree. 
I was limiting my self to reading from a local file and didn't see a big picture with network failures and chunked messages.
Thanks for pointing it out.

serii833 replied:
I guess adding a distinguished 'end of message' to my log messages would be a proper solution.

arnecls replied:
I actually like the idea of looking for the start of a message instead of the end. In some cases you cannot change the log messages and the start of a message most of the time is a quite good indicator. However - dealing with dangling messages is a bit tricky, but not impossible.

That's why I put both methods in the change request serii833#1, so that e.g. gollum can allow both types (and use timeouts to catch dangling messages)

arnecls committed:
Change multiline reading to regex with optional multiline support.
- Rename multiline to regex.
- Split into two modes: find start, find end.
- Leave unmatched data in the buffer to match behavior of other modes.
- Add a possibility to get incomplete data from the buffer.

arnecls committed:
fix incomplete message regex parsing

arnecls committed:
add more regex unittests and fix errors found.
- Proper handling of the incomplete flag during reset
- Add a function to check if there is incomplete data
- Fix inclusion of regex delemiter in all cases

arnecls commented:
Improvements to this PR have been committed to https://github.com/serii833/tgo/pull/1

arnecls committed:
remove debug output

arnecls committed:
clarify parsing comments

arnecls committed:
simplify loop body

serii833 committed:
Merge pull request #1 from trivago/pr-9-refinement

Proposal to change PR#9 to trivago tgo

arnecls commented:
Thank you for this and the good collaboration 👍

serii833 commented:
Thanks for your help :smile:
